### PR TITLE
modularize reinforcement learning for DA model

### PR DESCRIPTION
**Pull request by cz9779**
This PR fixes [#305 ]. 

### Description of changes
Create a data augmentation wrapper `MetaAugmentationWrapper` that trains the given data augmentation model (parameter phi) by maximizing a given classifier's performance on the validation set.
Create a `MetaModel` that copies the classifier and update its parameters.
An example is given in `examples/da_rl` with Bert MaskedLM data augmentation model and Bert classifier model.

### Test Conducted
Experiments with this example model are conducted.

### CODE REVIEW

File: examples/da_rl/download_imdb.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+# limitations under the License.
+
+import os
+import sys
+import subprocess
+
+
+def main():
+    if not os.path.exists("data/IMDB_raw"):
+        subprocess.run("mkdir data/IMDB_raw", shell=True, check=True)
```
Comment (hunterhector) [inline examples/da_rl/download_imdb.py:22]: Use the python one "https://github.com/asyml/forte/blob/master/forte/utils/utils_io.py#L28", don't use Linux commands because users may not work on a linux.

File: examples/da_rl/download_imdb.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    # pylint: disable=line-too-long
+    subprocess.run(
+        'wget -P data/IMDB_raw/ https://github.com/google-research/uda/blob/master/text/data/IMDB_raw/train_id_list.txt',
```
Comment (hunterhector) [inline examples/da_rl/download_imdb.py:25]: https://github.com/asyml/forte/blob/master/forte/data/data_utils.py#L50

Use this, not `wget`

File: examples/da_rl/run.sh [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# under examples/da_rl dir
+# prepare data
+python download_imdb.py
+python utils/imdb_format.py --raw_data_dir=data/IMDB_raw/aclImdb --train_id_path=data/IMDB_raw/train_id_list.txt --output_dir=data/IMDB
+python prepare_dataset.py
+# run model
+python main.py --do-train --do-eval --do-test
```
Comment (hunterhector) [inline examples/da_rl/run.sh:7]: new line at the end.

File: examples/da_rl/utils/data_utils.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
+# coding=utf-8
```
Comment (hunterhector) [inline examples/da_rl/utils/data_utils.py:1]: why utf-8?

File: forte/models/da_rl/generator_trainer.py [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+        aug_tokenizer:
+            A Bert-based tokenizer.
+        device:
+            The CUDA device to run the model on.
+        num_aug:
+            The number of samples from the LM for an augmented training example.
+            See :meth:`_augment_example` for implementation details.
+    """
+
+    def __init__(self, augmentation_model, augmentation_optimizer,
```
Comment (hunterhector) [inline forte/models/da_rl/generator_trainer.py:62]: Type annotation missing here.

File: forte/models/da_rl/generator_trainer.py [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+        self._aug_optimizer = augmentation_optimizer
+        self._aug_tokenizer = aug_tokenizer
+        self._device = device
+        self._num_aug = num_aug
+
+    def reset_model(self):
+        self._aug_model.train()
+        self._aug_model.zero_grad()
+
+    def _augment_example(self, features, num_aug):
```
Comment (hunterhector) [inline forte/models/da_rl/generator_trainer.py:74]: let's use `instance` instead of `example`

File: forte/models/da_rl/generator_trainer.py [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+
+        len = int(torch.sum(input_mask).item())
+        if len >= 4:
+            mask_idx = sorted(
+                random.sample(list(range(1, len - 1)), max(len // 7, 2)))
+        else:
+            mask_idx = [1]
+
+        # init_ids[0][mask_idx] = \
+        #     self._aug_tokenizer.convert_tokens_to_ids(['[MASK]'])[0]
```
Comment (hunterhector) [inline forte/models/da_rl/generator_trainer.py:88]: Remove redundant commented lines.

File: forte/models/da_rl/generator_trainer.py (hunk 6 above)
Comment (hunterhector) [inline forte/models/da_rl/generator_trainer.py:74]: For an important function like this, document clearly what this do. What are the input? what are the outputs? what are the dimensions?

File: forte/models/da_rl/generator_trainer.py [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+            [ap.unsqueeze(0) for ap in aug_probs_all], axis=0)
+
+        return aug_probs
+
+    def augment_example(self, features):
+        r"""Augment a training example.
+
+        Args:
+            features: A tuple of Bert features of one training example.
+                (input_ids, input_mask, segment_ids, label_ids).
```
Comment (hunterhector) [inline forte/models/da_rl/generator_trainer.py:120]: Dimensions of these tensors?

File: forte/models/da_rl/generator_trainer.py [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+
+            if group['weight_decay'] != 0:
+                grad = grad + group['weight_decay'] * param.data
+
+            exp_avg = exp_avg * beta1 + (1. - beta1) * grad
+            exp_avg_sq = exp_avg_sq * beta2 + (1. - beta2) * grad * grad
+            denom = exp_avg_sq.sqrt() + group['eps']
+
+            # no bias correction
+            # bias_correction1 = 1. - beta1 ** step
```
Comment (hunterhector) [inline forte/models/da_rl/generator_trainer.py:275]: Are these lines to be deleted?

File: forte/models/da_rl/magic_model.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+    @property
+    def _flat_weights(self):
+        return [p for layerparams in self.all_weights for p in layerparams]
+
+    @property
+    def all_weights(self):
+        return [[getattr(self, weight) for weight in weights] for weights in
+                self._all_weights]
+
+    # https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/container.py#L150
```
Comment (hunterhector) [inline forte/models/da_rl/magic_model.py:111]: What do these comments mean?

File: examples/da_rl/utils/data_utils.py [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This is the Data Loading Pipeline for Sentence Classifier Task from:
+    `https://github.com/google-research/bert/blob/master/run_classifier.py`
+"""
+
```
Comment (hunterhector) [inline examples/da_rl/utils/data_utils.py:19]: Looks like you are using the whole Google system for classifier. We need to use the Forte + Texar implementation instead.

**Review by hunterhector — CHANGES_REQUESTED** [review]
1. It looks like the classification module is not done using Forte (I didn't see a data reader) nor Texar (I didn't see a Texar module). This example does not look like it belongs to Forte. For example, you are using `imdb_format.py`, which is totally unrelated to Forte format.
2. Type annotations are all missing, please add.
3.  We need a Readme to explain this.
4.  The `main` function is too long for people to locate relevant information.
5. We need test cases for the modules

File: forte/models/da_rl/magic_model.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+        assert num_heads == self._hparams.num_heads
+        return torch.reshape(t, (t.size(0), t.size(1), num_heads * dim))
+
+    def transpose_for_scores(self, x):
+        new_x_shape = x.size()[:-1] + (
+            self.num_attention_heads, self.attention_head_size)
+        x = x.view(*new_x_shape)
+        return x.permute(0, 2, 1, 3)
+
+    def conv2d_forward(self, input, weight):
```
Comment (hunterhector) [inline forte/models/da_rl/magic_model.py:175]: why do we have convolution here?

File: forte/models/da_rl/generator_trainer.py [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+
+    theta'(phi) = theta - \nabla_{theta} L_{train}(theta, phi)
+
+    phi = phi - \nabla_{phi} L_{val}(theta'(phi))
+
+    Args:
+        augmentation_model:
+            A Bert-based language model for data augmentation.
+        augmentation_optimizer:
+            An optimizer.
```
Comment (hunterhector) [inline forte/models/da_rl/generator_trainer.py:52]: These comments do not provide additional information than the name itself, can you elaborate?

**Review by hunterhector — COMMENTED** [review]
Are we going to remove these `da_rl` directories? IN this PR we should focus on the core reinforcement learning part.
